RAINCATCH-1108 Datasync cloud and client

austincunningham opened the pull request:
## Motivation
document the data sync  cloud and client modules 

## What ?
document Datasync cloud and client usage in the demo application and how to implement as a developer. 

- [ ] Documentation
- [ ] gh-pages


@wtrocki @austincunningham 
@jstaffor docs changes to this PR https://github.com/austincunningham/raincatcher-docs/pull/4

austincunningham committed:
initial commit

austincunningham committed:
update

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+= {WFM-RC-NameLong} Datasync Module
+
+As with all synchronisation tasks Datasync Module has cloud and client sides. This is a implementation of 
+the Feedhenry Sync Framework for advanced usage see the 
```
From a downstream POV, "Feedhenry" should be avoided. Should FeedHenry be a variable?

jstaffor replied:
"Sync Framework. For advanced usage,  see the "

austincunningham replied:
agree

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+As with all synchronisation tasks Datasync Module has cloud and client sides. This is a implementation of 
+the Feedhenry Sync Framework for advanced usage see the 
+link:{WFM-RC-SyncFrameworkURL}{WFM-RC-Branch}/docs[sync framework documentation]. 
+The primary funtion of Datasync is to synchronise mobile data between mobile and a backend server.
```
Suggestion:
"The primary function the *Datasync* module is to synchronise data between the client (mobile) and a backend server (cloud)." - 'cloud' may be wrong in this instance.

jstaffor replied:
"Client Apps" and "Cloud Apps" mentioned below so maybe "a Client App, and a Cloud App running on a backend server"

What very important here is that whatever the decision is, it's reflected throughout this PR so there is consistency.

austincunningham replied:
````The primary funtion of Datasync module is to synchronise data between mobile(Client App) and a backend server(Cloud App).````

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+The primary funtion of Datasync is to synchronise mobile data between mobile and a backend server.
+The key features of Datasync are as follows:
+
+- Allows mobile apps to use and update data offline (local cache)
```
"update data in the local cache, offline"

austincunningham replied:
bullet points from Sync documentation.  Will not change as will change the meaning.

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+The key features of Datasync are as follows:
+
+- Allows mobile apps to use and update data offline (local cache)
+- Provides a mechanism to manage bi-directional data synchronization from multiple Client Apps via the Cloud App and into back-end data stores
```
What's this mechanism? It would be good to have the name of this in brackets after the word "mechanism".

austincunningham replied:
same as last comment

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+
+- Allows mobile apps to use and update data offline (local cache)
+- Provides a mechanism to manage bi-directional data synchronization from multiple Client Apps via the Cloud App and into back-end data stores
+- Allows data updates (that is, deltas) to be distributed from the Cloud App to connected clients
```
My understanding of the word "delta" is the difference(and *only* the difference) between data in a given set A, and a given set B. This line needs to reflect this as this is not clear from reading it IMO.

austincunningham replied:
same as last comment

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+- Allows mobile apps to use and update data offline (local cache)
+- Provides a mechanism to manage bi-directional data synchronization from multiple Client Apps via the Cloud App and into back-end data stores
+- Allows data updates (that is, deltas) to be distributed from the Cloud App to connected clients
+- Provides facilities for managing data collisions from multiple updates in the cloud
```
"data collisions" - what exactly is this? Suggestion: "Provides facilities for managing <brief explanation of data collisions> when this occurs across multiple updates in the cloud.

or

Leave what is there and add "For more information on data collisions, please see <appropriate link>"

austincunningham replied:
same as above

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+- Allows data updates (that is, deltas) to be distributed from the Cloud App to connected clients
+- Provides facilities for managing data collisions from multiple updates in the cloud
+
+This service effectively allows Sync Apps to seamlessly continue working when the network connection is lost, and allow them to recover when the network connection is restored.
```
The *Datasync* service eff...

"The *Datasync* service allows Sync Apps to continue working seamlessly when the network connection is lost..."

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+This service effectively allows Sync Apps to seamlessly continue working when the network connection is lost, and allow them to recover when the network connection is restored.
+
+== Requirements
+Our demo application has the following requirements for Datasync Cloud 
```
The <WFM Variable> demo application...

austincunningham replied:
:+1:

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+
+
+== {WFM-RC-NameLong} Datasync cloud
+Datasync cloud has to be used in conjunction with the Datasync client/mobile applications.
```
"Datasync cloud must be used"

"Datasync mobile (Client App) applications."

austincunningham replied:
"Datasync cloud must be used"
:+1: 
"Datasync mobile (Client App) applications."
:-1:  module name is Datasync Client, will remove mobile if confusing

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+
+== {WFM-RC-NameLong} Datasync cloud
+Datasync cloud has to be used in conjunction with the Datasync client/mobile applications.
+{WFM-RC-NameLong} Datasync cloud is a wrapper for the Feedhenry Datasync server.
```
From a downstream POV, "Feedhenry" should be avoided. Should FeedHenry be a variable?

austincunningham replied:
replaced with link to repo.

austincunningham replied:
removed

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+== {WFM-RC-NameLong} Datasync cloud
+Datasync cloud has to be used in conjunction with the Datasync client/mobile applications.
+{WFM-RC-NameLong} Datasync cloud is a wrapper for the Feedhenry Datasync server.
+This module creates server side service that allows mobile applications to sync offline data to preferred database.
```
"The *Datasync* module creates a server side service..

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+Datasync cloud has to be used in conjunction with the Datasync client/mobile applications.
+{WFM-RC-NameLong} Datasync cloud is a wrapper for the Feedhenry Datasync server.
+This module creates server side service that allows mobile applications to sync offline data to preferred database.
+By default server connects and stores all client data in mongodb database. 
```
"By default, the server "

Also, what server, the back end server?

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+By default server connects and stores all client data in mongodb database. 
+
+
+=== Implementation of {WFM-RC-NameLong} Datasync cloud
```
It is V IMPORTANT we get this clear to avoid the same sit we have with Passport.js/PassportAuth. 

"Datasync Module" was mentioned above, now a concept of "Datasync cloud" is being implemented....should it be "Datasync Module"? If not, I think it is very important that we distinguish between "Datasync Module" and "Datasync cloud" at the start of this adoc. If you want to talk this through - do ping me and lets talk it through.

austincunningham replied:
Datasync module has a cloud module and a client module. wtrocki felt it should be documented as one module, as they don't work independently

austincunningham replied:
Added the following line to introduction.

```The cloud is implemented with the Datasync cloud module and client is implemented with the Datasync client module. ```

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+
+This service effectively allows Sync Apps to seamlessly continue working when the network connection is lost, and allow them to recover when the network connection is restored.
+
+== Requirements
```
Prerequisites? (think about consistency with merged docs)

austincunningham replied:
ok

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+
+=== Implementation of {WFM-RC-NameLong} Datasync cloud
+
+Datasync Cloud is used in conjunction with an Express Server. 
```
This is not called out in the Requirements section

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+Datasync Cloud is used in conjunction with an Express Server. 
+
+
+== {WFM-RC-NameLong} Datasync client 
```
Again, it is V IMPORTANT we get this clear to avoid the same sit we have with Passport.js/PassportAuth.

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+
+
+== {WFM-RC-NameLong} Datasync client 
+Datasync client requres a server running Datasync Cloud.  Data synchonization Library can be used for offline storage 
```
" a server " - a Node.js server? What type? 

Also, it this called out in the Requirements?

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+
+== {WFM-RC-NameLong} Datasync client 
+Datasync client requres a server running Datasync Cloud.  Data synchonization Library can be used for offline storage 
+of mobile applications data.
```
"mobile (Client App) applications data." - only a suggestion given what I mentioned above.

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-cloud-module.adoc:
```diff
@@ -0,0 +1,42 @@
+= {WFM-RC-NameLong} Datasync Module
```
should the heading be {WFM-RC-NameLong} Datasync Cloud given that the name of the adoc is "con-datasync-cloud-module.adoc"?

austincunningham replied:
renamed the file

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-module.adoc:
```diff
@@ -0,0 +1,42 @@
+= {WFM-RC-NameLong} Datasync Module
```
I think this file was not to be added? After a brief look, it looks identical to the text in the adoc above.

austincunningham replied:
the rename left the original file. removed the original

jstaffor reviewed (commented):
Feedback resulting from initial observations of this PR.

austincunningham committed:
addressing pr comments

austincunningham committed:
update documentation

austincunningham committed:
merged master

austincunningham committed:
merged master

austincunningham committed:
merged master

austincunningham committed:
addressing wtrocki comments

wtrocki commented:
Content is ok, but we are really light in terms of actual purpose etc. We can try to be more RainCatcher specific in next PR's

jstaffor commented on docs/workforce-management-framework/topics/con-datasync-module.adoc:
```diff
@@ -0,0 +1,39 @@
+The cloud is implemented with the Datasync cloud module and client is implemented with the Datasync client module. 
+These are a implementation of the Sync Framework for advanced usage see the 
+link:{WFM-RC-SyncFrameworkURL}{WFM-RC-Branch}/docs[sync framework documentation]. 
+The primary funtion of Datasync is to synchronise mobile data between mobile(client) and a backend server(cloud).
```
@wtrocki - are you ok with the update of terms - basically I'm changing this:
 - "The primary funtion of Datasync is to synchronise mobile data between mobile(client) and a backend server(cloud)."

to this:

- "The primary function of *Datasync* is to synchronize data between the Client Apps (mobile) and the Cloud App (running on a backend server)."

It's more inline with our existing documentation - ie: the identification of a Client App (mobile) and a Cloud App (running on a back end server).

NOTE: It'll have a knock on affect to the rest of the doc that I'll handle.

witmicko commented:
Closing in favour of #47

jstaffor committed:
RAINCATCH-1108 - changes made and ready for tech approval

jstaffor commented:
@austincunningham, @witmicko - changes made in (https://github.com/austincunningham/raincatcher-docs/pull/4)

austincunningham committed:
Merge pull request #4 from jstaffor/RAINCATCH-1108

RAINCATCH-1108 - changes made and ready for tech approval

austincunningham committed:
merged master

austincunningham committed:
Merge branch 'RAINCATCH-1108' of github.com:austincunningham/raincatcher-docs into RAINCATCH-1108

jstaffor committed:
Unrelated changes to attributes.adoc removed 

Removed the following unrelated lines:

:WFM-RC-KeycloakURL: https://keycloak.gitbooks.io/documentation/
:WFM-RC-AngularJsURL: https://github.com/feedhenry-raincatcher/raincatcher-angularjs/blob/
:WFM-RC-CoreURL: https://github.com/feedhenry-raincatcher/raincatcher-core/blob/
:WFM-RC-KeycloakDockerURL: https://github.com/feedhenry-raincatcher/raincatcher-keycloak/blob/

jstaffor commented:
@austincunningham - you might want to take a note of these so they are not lost:

https://github.com/feedhenry-raincatcher/raincatcher-docs/pull/41/commits/227454d07d2d5d7ad88780d3c8efc9291acc33d6

austincunningham committed:
merged conflicts
